In [1]:
import pandas as pd
import numpy as np
from itables import show

In [2]:
df = pd.read_parquet('../data/spend_data_large_sourcing.parquet')
show(df)
df.shape

(698411, 51)

## Remove Unneeded Columns
### Needed Columns
1. Spend_Data_Posting_Date_Fiscal_Year
2. Spend_Data_Posting_Date_Fiscal_Year_and_Quarter
3. Spend_Data_Posting_Date
4. Spend_Data_Vendor_Invoice_Date
5. Spend_Data_PO_Number
6. Spend_Data_PO_Line_Item_Number
7. Spend_Data_Vendor_Number
8. Spend_Data_Invoice_Type
9. Spend_Data_Business_Partner_Name_
10. Spend_Data_Organization
11. Spend_Data_Division_Description
12. Spend_Data_Business_Unit_Description
13. Spend_Data_Geo
14. Spend_Data_Line_Item_Text
15. Spend_Data_Vendor_Invoice_Amount_LC2_USD_
16. Spend_Data_Vendor_Name
17. Spend_Data_Sourcing_Category
18. Spend_Data_Sourcing_Portfolio

## Filter Down Spend Data - Starting Point (698,411 rows; 51 columns)
1. Remove rows where Spend_Data_Invoice_Type is 'NonPO-Related'
2. Spend_Data_Sourcing_Portfolio = 'Technology


In [3]:
#Options available in Spend_Data_Sourcing_Category
df['Spend_Data_Sourcing_Category'].unique()

array(['Legal', 'Branding', 'Non-addressable', 'Advertising', 'Financial',
       'Enterprise managed software', 'Talent acquisition contingent',
       'Network and telecom', 'Facility management', 'Office products',
       'Premises', 'Technology professional services', 'Web production',
       'Benefits', 'Business process outsourcing', 'Infrastructure',
       'Gifts & Awards', 'Business advisory services', 'Localization',
       'Teleservices', 'Training', 'Events', 'Talent acquisition perm',
       'Global Mobility', 'Hardware', 'Reseller', 'Royalties',
       'HR consulting', 'Market research', 'License compliance', 'Impact',
       'Ground Transport'], dtype=object)

In [4]:
#1 remove rows where Spend_Data_Invoice_Type is 'NonPO-Related'
df = df[df['Spend_Data_Invoice_Type'] != 'NonPO-Related']
print(df.shape)

#2 remove rows where Spend_Data_Sourcing_Portfolio is not 'Technology'
df = df[df['Spend_Data_Sourcing_Portfolio'] == 'Technology']
print(df.shape)

#3 filter down to Enterprise sourcing category
df = df[df['Spend_Data_Sourcing_Category'] == 'Enterprise managed software']
print(df.shape)


(390541, 51)
(76136, 51)
(8082, 51)


In [5]:
#count rows in each fiscal year
df['Spend_Data_Posting_Date_Fiscal_Year'].value_counts()

Spend_Data_Posting_Date_Fiscal_Year
2024    2191
2025    2166
2023    2160
2026    1565
Name: count, dtype: int64

In [6]:
#filter for 24,25,26
df = df[df['Spend_Data_Posting_Date_Fiscal_Year'].isin([2024,2025,2026])]
print(df.shape)


(5922, 51)


In [7]:
#Count unique vaules with (Vendor_Name, Spend_Data_Sourcing_Category)
print("Number of unique values with (Vendor_Number, line item text):")
df.groupby(['Spend_Data_Vendor_Number', 'Spend_Data_Line_Item_Text']).ngroups

Number of unique values with (Vendor_Number, line item text):


3226

### Count cost to run the model on the data

In [8]:
from tokens import calculate_exact_ai_cost

# Define which columns you will actually send to the AI
cols_to_use = [
    'Spend_Data_Vendor_Name', 
    'Spend_Data_Line_Item_Text', 
    'Spend_Data_GL_Account_Description' # Add others if you use them
]

# Run the calculation on your filtered df
calculate_exact_ai_cost(df, cols_to_use, model="gpt-4o-mini")

Counting tokens per row...

--- Exact Cost Analysis for 5,922 rows (gpt-4o-mini) ---
Columns used: ['Spend_Data_Vendor_Name', 'Spend_Data_Line_Item_Text', 'Spend_Data_GL_Account_Description']
Total Input Tokens:  254,312
Avg Tokens/Row:      42.9
Max Tokens/Row:      75
Est. Total Cost:     $0.2158


np.float64(0.21580679999999997)

In [9]:
# Generate a lookup table for unqiue (vendor_number, line_item_text)
lookup_table = df.groupby(['Spend_Data_Vendor_Name', 'Spend_Data_Vendor_Number', 'Spend_Data_Line_Item_Text']).size().reset_index(name='count')

#add columns parent and sub_category
lookup_table['category'] = None
lookup_table['sub_category'] = None

show(lookup_table)

print(len(lookup_table))

3226


In [11]:
from functions import categorize_lookup_table, retry_errors

# Test with 20 rows first (2 API calls)
lookup_table = categorize_lookup_table(lookup_table, batch_size=None, rows_per_call=10, delay=0.2)

# View results
show(lookup_table)

Getting access token...
✓ Token received

Processing 3206 rows in batches of 10...
Estimated API calls: 321


Batches:   3%|▎         | 9/321 [00:34<18:36,  3.58s/it]

💾 Checkpoint saved (99 rows processed)


Batches:   6%|▌         | 19/321 [01:03<14:39,  2.91s/it]

💾 Checkpoint saved (198 rows processed)


Batches:   9%|▉         | 29/321 [01:34<14:21,  2.95s/it]

💾 Checkpoint saved (298 rows processed)


Batches:  12%|█▏        | 39/321 [02:01<12:23,  2.64s/it]

💾 Checkpoint saved (398 rows processed)


Batches:  15%|█▌        | 49/321 [02:39<16:51,  3.72s/it]

💾 Checkpoint saved (498 rows processed)


Batches:  18%|█▊        | 59/321 [03:12<14:10,  3.25s/it]

💾 Checkpoint saved (598 rows processed)


Batches:  21%|██▏       | 69/321 [03:43<12:49,  3.05s/it]

💾 Checkpoint saved (698 rows processed)


Batches:  25%|██▍       | 79/321 [04:17<13:32,  3.36s/it]

💾 Checkpoint saved (798 rows processed)


Batches:  28%|██▊       | 89/321 [04:54<13:58,  3.61s/it]

💾 Checkpoint saved (898 rows processed)


Batches:  31%|███       | 99/321 [05:30<13:52,  3.75s/it]

💾 Checkpoint saved (998 rows processed)


Batches:  34%|███▍      | 109/321 [06:09<13:44,  3.89s/it]

💾 Checkpoint saved (1097 rows processed)


Batches:  37%|███▋      | 119/321 [06:45<11:44,  3.49s/it]

💾 Checkpoint saved (1197 rows processed)


Batches:  40%|████      | 129/321 [07:24<12:34,  3.93s/it]

💾 Checkpoint saved (1297 rows processed)


Batches:  43%|████▎     | 139/321 [07:55<09:00,  2.97s/it]

💾 Checkpoint saved (1397 rows processed)


Batches:  46%|████▋     | 149/321 [08:33<10:32,  3.67s/it]

💾 Checkpoint saved (1497 rows processed)


Batches:  50%|████▉     | 159/321 [09:14<11:01,  4.08s/it]

💾 Checkpoint saved (1596 rows processed)


Batches:  53%|█████▎    | 169/321 [09:46<08:30,  3.36s/it]

💾 Checkpoint saved (1695 rows processed)


Batches:  56%|█████▌    | 179/321 [10:17<06:48,  2.88s/it]

💾 Checkpoint saved (1795 rows processed)


Batches:  59%|█████▉    | 189/321 [10:46<06:15,  2.85s/it]

💾 Checkpoint saved (1894 rows processed)


Batches:  62%|██████▏   | 199/321 [11:15<05:46,  2.84s/it]

💾 Checkpoint saved (1994 rows processed)


Batches:  65%|██████▌   | 209/321 [11:46<05:17,  2.84s/it]

💾 Checkpoint saved (2093 rows processed)


Batches:  68%|██████▊   | 219/321 [12:14<04:56,  2.91s/it]

💾 Checkpoint saved (2193 rows processed)


Batches:  71%|███████▏  | 229/321 [12:43<04:36,  3.00s/it]

💾 Checkpoint saved (2293 rows processed)


Batches:  74%|███████▍  | 239/321 [13:16<04:34,  3.35s/it]

💾 Checkpoint saved (2393 rows processed)


Batches:  78%|███████▊  | 249/321 [13:55<05:26,  4.53s/it]

💾 Checkpoint saved (2493 rows processed)


Batches:  81%|████████  | 259/321 [14:28<03:34,  3.46s/it]

💾 Checkpoint saved (2593 rows processed)


Batches:  84%|████████▍ | 269/321 [15:07<03:29,  4.03s/it]

💾 Checkpoint saved (2693 rows processed)


Batches:  87%|████████▋ | 279/321 [15:39<02:04,  2.98s/it]

💾 Checkpoint saved (2793 rows processed)


Batches:  90%|█████████ | 289/321 [16:12<01:41,  3.17s/it]

💾 Checkpoint saved (2893 rows processed)


Batches:  93%|█████████▎| 299/321 [16:41<01:02,  2.84s/it]

💾 Checkpoint saved (2993 rows processed)


Batches:  96%|█████████▋| 309/321 [17:14<00:42,  3.53s/it]

💾 Checkpoint saved (3093 rows processed)


Batches:  99%|█████████▉| 319/321 [17:42<00:05,  2.83s/it]

💾 Checkpoint saved (3192 rows processed)


Batches: 100%|██████████| 321/321 [17:44<00:00,  3.32s/it]


✓ PROCESSING COMPLETE
📊 Results:
   Rows processed:  3,198
   Errors:          0

📈 Token Usage:
   Input tokens:    403,982
   Output tokens:   77,983
   Total tokens:    481,965

💰 Cost (GPT-4o-mini pricing):
   Input cost:      $0.0606
   Output cost:     $0.0468
   Total cost:      $0.1074

💾 Checkpoint saved to: lookup_checkpoint.parquet


## Lets match categories and subcategories to df from lookup table

step 1: add columns to df for category and sub_category

step 2: iterate through df
    - for each line see if the vendor_number and line_item_text are in the lookup table
    - if so, add the category and sub_category to the df
    - if not, add a null value to the category and sub_category

step 3: return df with category and sub_category columns



In [12]:
#make a copy of df just to be safe
df_copy = df.copy()

In [13]:
#Lets match categories and subcategories to df from lookup table

# Ensure key columns match types (e.g., both strings)
df['Spend_Data_Vendor_Number'] = df['Spend_Data_Vendor_Number'].astype(str)
df['Spend_Data_Line_Item_Text'] = df['Spend_Data_Line_Item_Text'].astype(str)

lookup_table['Spend_Data_Vendor_Number'] = lookup_table['Spend_Data_Vendor_Number'].astype(str)
lookup_table['Spend_Data_Line_Item_Text'] = lookup_table['Spend_Data_Line_Item_Text'].astype(str)



In [14]:
# Merge category and sub_category back to main df
df = df.merge(
    lookup_table[['Spend_Data_Vendor_Number', 'Spend_Data_Line_Item_Text', 'category', 'sub_category']],
    on=['Spend_Data_Vendor_Number', 'Spend_Data_Line_Item_Text'],
    how='left'
)

In [16]:
show(df)

print(df.shape)

(5922, 53)


In [17]:

# Save to data folder
df.to_parquet('../data/spend_data_categorized.parquet', index=False)

In [18]:
df.to_excel('../data/spend_data_categorized.xlsx', index=False)

## use lookup table on steves data

In [19]:

df_spend_new = pd.read_csv('../data/FY23-FY26 Enterprise Managed Software Spend.csv')

#add columns for category and sub_category
df_spend_new['category'] = None
df_spend_new['sub_category'] = None


In [20]:
#merge
df_spend_new = df_spend_new.merge(
    lookup_table[['Spend_Data_Vendor_Number', 'Spend_Data_Line_Item_Text', 'category', 'sub_category']],
    left_on=['Vendor Number', 'Line Item Text'],
    right_on=['Spend_Data_Vendor_Number', 'Spend_Data_Line_Item_Text'],
    how='left'
)

show(df_spend_new)
